### ISSUE TITLE
Add /title command

### ISSUE THREAD

**Issue by sphinxc0re** [cuberite/cuberite#1806]
I only found the german wiki page for this (http://minecraft-de.gamepedia.com/Befehl/title) I hope you don't mind.
Also:
http://wiki.vg/Protocol#Title

**Comment by madmaxoft** [comment]
I've seen this used in the Plots plugin on some servers, it is quite a useful feature - automatically listing the plot owners, or other useful information, without having to query.

This should be rather easy to add:
- add the packet sending function(s) in the protocol class (need to research about when the packet was added; implement as NOOP for protocols that don't support it)
- add the packet sending function(s) to a `cClientHandle` class 
- export the `cClientHandle` functions to Lua (should be possible to do this automatically)

I think it'd be better to implement the interface as multiple functions rather than one single one:
```cpp
void SendSetTitle(const cCompositeChat & a_Title, const cCompositeChat & a_SubTitle);
void SendSetTitle(const AString & a_Title, const AString & a_SubTitle);
void SendTitleTimes(int a_FadeInTicks, int a_DisplayTicks, int a_FadeOutTicks);
void SendHideTitle();
void SendResetTitle();
```

**Comment by flx5** [comment]
It was added with 14w20a (http://wiki.vg/Protocol_History#14w20a) which equals protocol version 18 (http://wiki.vg/Protocol_version_numbers).

**Comment by scottwillmoore** [comment]
I decided to attempt working on this feature as it appears fairly simple to implement and you have given me all the neccesary steps to implement it! I do not have much experience working with C++, however I am fairly experienced in Java and have a fairly solid understanding of the Minecraft internals as I have explored much of the Minecraft source code.

So far I have implemented the functions you suggested above in the base `cProtocol` class, then I have also overriden the functions in the implementing headers `cProtocol172` and `cProtocol180` and have now began implementing the functionality. As mentioned above it was implemented in a 1.8 snapshot and is therefore not avaliable in the 1.7 released, so I have implemented a NOOP for each of the new functions.

I have read through your source code and the implementation appears fairly simple I just have a few simple questions. 

Firstly when setting up the environment in the `GETTING-STARTED` guide it mentions to run `src/Bindings/AllToLua.bat`, however it appears I do not have the `tolua++` executable? I have followed the instructions so should I manually install the `tolua++` or is it supposed to build with the files. I ended up skipping this step for the time being.

Next when implementing the functions I noticed you have several different types of data encodings which you can choose from such as `WriteBEInt8` vs `WriteBEInt16` and I do not know which one I should be choosing or whether it matters? The protocols mentioned on the wiki do not tell us what size the data should be?

Also while looking at the [wiki](http://wiki.vg/Protocol#Title) it appears sending the title is sent in a separate packet to sending the subtitle. So perhaps the functions need to be changed. It is sent in a similar way to how the scoreboard information is sent. The functions may need changing to the following, or perhaps sending two packets in a single function? However you feel will work best.
````c++
void SendSetTitle(const cCompositeChat & a_Title)
void SendSetTitle(const AString & a_Title)
void SendSetSubTitle(const cCompositeChat & a_SubTitle)
void SendSetSubTitle(const AString & a_SubTitle)
````

**Comment by madmaxoft** [comment]
Hello, and welcome to the team :)

I suppose the GEETTING-STARTED guide is outdated on this, you don't need to run any script prior to building. Just use cmake and then build using your favorite tools (make / MSVC IDE / ...). Let me know (PM me on the forum for best results) if this doesn't work.

The datatypes are a vital part of the protocol - if you send BEInt16 and the client expects only a single-byte value, things won't work. The wiki actually does say what the datatypes are - it's in the "Field type" column, although you need to translate those. "int", for example, is either `BEInt32` or `BEUInt32`, "VarInt" is either `VarInt32` or `VarInt64` (doesn't really matter, the encoding for both sizes is the same, it's only about the range), "Chat" is a simple `String` in the end.

I proposed the function signatures for easier usage - you want to set both the title and the subtitle at once, so you can just call one function and have it send two packets. That's perfectly reasonable, the `cProtocol` class should be an abstraction over the protocol, not a 1:1 mapping for the packets, but rather an interface with the "logical calls" as units.
Of course we're open to suggestions, if you think sending a subtitle only has its merits, then by all means share your thoughts on this.

**Comment by scottwillmoore** [comment]
Thanks for the quick response :)

Alright so the datatypes arn't specified in the protocol, therefore it must be translated into an acutal value. Where `Int -> BEInt32 or BEUInt32`, `VarInt -> BEInt64, BEUInt64`, `Chat -> String`, etc. I also assume there is no difference between an `Int32` and a `UInt32`, but just in the way they are read, they are both comprised of 16 bits, but `Int32` uses one of these bits to determine the sign. Also would `Int8` just be a byte just be represented by an `Int8`?

It also does make sense to combine both the title and subtitle into one function, however it may be better to leave them separate as they are acutally separate commands ([detailed here](http://minecraft.gamepedia.com/Commands?cookieSetup=true#title)). This may make it easier to implement the core plugin later?

Read the effects of each command and I believe now they must be separate functions, as calling the `/title <player> title ...` command displays the title as well as setting the title text, while calling `/title <player> subtitle ...` only sets the subtitle for the next time that the title command is called, it dos not call a render for the title on the client. Based on this I think it may be neccessary to implement them separately?

**Comment by madmaxoft** [comment]
"VarInt" is not `BEInt64` nor `BEUInt64`. "VarInt" is a variable-length encoding for an unsigned integer, `BEInt64` is a fixed-length 64-bit integer and `BEUInt64` is a fixed length 64-bit unsigned integer. `WriteVarInt32` is the proper MCS call to write a "VarInt" to the packet buffer.

`Int32` is 32-bits long, not 16-bits. `Int8` is a single byte, yes.

Those commands make it clear that the API needs to be split, true.

And I see that I've been mistaken, I thought the title was something completely else, I guess what I've seen was the chat message in different positions instead. Doesn't matter, as long as we gain a new feature :)

**Comment by scottwillmoore** [comment]
Oh yeah, my mistake I meant 32 bits not 16. Thanks for the info.

I have split the commands an implemented them in the cProtocol class as well as all deriving classes. I have also created functions that replicate that behaviour in the cClientHandle as well, it all appears to compile and work well, but now a lua implementation of the command must be created.

I have read through many of the other lua plugins and managed to get my own lua plugin running (also the webadmin interface is really interesting btw), but I cannot figure out how to use the functions I have created from within the lua file. The title command requires you to specify a player (or group of players when using @a, etc), which will be handled within the lua, but how do I get that player and then access the cClientHandle interface in order to send the packet to that player. Do I need to add it to the lua bindings or should it already be generated after building the source code?
